client/src/components/game/hooks/gameConnection/UseGameConnection.tsx

```diff
@@ -0,0 +1,106 @@
+import { useState, useEffect, useReducer } from "react";
+import {
+  AddComment,
+  AlertPayload,
+  BatchPayload,
+  lobbyCode,
+  RequestType,
+  ResponsePayload,
+  ResponseType,
+} from "../../../../Interface";
+import gameReducer from "../../reducers/GameReducer";
+import config from "../../../../config/Config";
+
+function getWebSocketAddress(sessionId: string, username: string) {
+  return `${config.wsUri}/${config.socket.uri
+    .replace(":sessionId", sessionId)
+    .replace(":username", username)}`;
+}
+
+function useGameConnection(
+  sessionId: string,
+  username: string,
+  onAlert: (alert: string) => void
+) {
+  const [ws, setWs] = useState<WebSocket>(null as unknown as WebSocket);
+  const [state, dispatch] = useReducer(gameReducer, {
+    players: [],
+    host: "",
+    source_code: lobbyCode,
+    comments: [],
+  });
+
+  const [disconnectionMessage, setDisconnectionMessage] = useState<
+    string | null
+  >(null);
+
+  useEffect(() => {
+    const socket = new WebSocket(getWebSocketAddress(sessionId, username));
+    socket.onmessage = ({ data }) => {
+      const payload: ResponsePayload = JSON.parse(data);
+      if (payload.message_type === ResponseType.ALERT) {
+        onAlert((payload as AlertPayload).alert);
+      } else if (payload.message_type === ResponseType.BATCH) {
+        const batchPayload = payload as BatchPayload;
+        const alertPayload = batchPayload.messages.find(
+          ({ message_type }) => message_type === ResponseType.ALERT
+        ) as AlertPayload;
+        if (alertPayload) {
+          onAlert(alertPayload.alert);
+        }
+      }
+
+      // AlertPayloads are ignored in the reducer
```

Comment:
I want to restrict the `GameReducer` to only performing state manipulations. `AlertPayload` messages only trigger toasts to display,  not alter the game state. This is why I ignored `AlertPayload` messages in the `GameReducer` and handled them directly above.

Reply:
Makes sense.